048-avoid-broker-restarts-when-in-recovery.md

```diff
@@ -4,17 +4,21 @@ This proposal describes a solution for KafkaRoller's shortcoming with brokers in
 
 ## Current situation
 
-The current logic in KafkaRoller:
+The current logic in KafkaRoller from high level:
 - Take a list of Kafka pods and reorder them based on the readiness.
-- Roll unready pods first without considering why it could be unready.
 - If the pod is the controller, roll it last.
-- Do not roll the pod, if it has an impact on the availability such as causing under replicated partitions.
-- If a Kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then it may force restart it without considering why it's taking a long time.
+- If pod is not ready before rolling, wait for it to become ready until the operational timeout reaches.
+- If pod is in a stuck state e.g CrashLoopBackOff, force restart the pod.
+- If pod is unschedulable, fail the reconcilation. 
+- If the AdminClient connection to pod failed, force restart the pod. 
+- Check if rolling the pod would have an impact on the availability such as causing under replicated partitions.
+- If rolling the pod does affect the availibility, then restart the pod and wait for pod readiness.
+- If pod does not became ready within the operational timeout (300000ms by default) after restarting, then retry until the maximum attempt is reached. 
 
 The current flow of KafkaRoller
 ![The current flow of KafkaRoller](images/048-kafka-roller-current-flow.png)
 
-A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller could end up force restarting the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+A Kafka broker can take a long time to become ready while performing log recovery. While broker is in log recovery, AdminClient connection cannot be made to the broker. In this case, KafkaRoller force restarts the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
```

Comment:
```suggestion
A Kafka broker can take a long time to become ready while performing log recovery. While the broker is in log recovery, AdminClient connection cannot be made to the broker. In this case, KafkaRoller force restarts the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
```